src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -669,6 +727,28 @@ public void ReportDiagnostic(Diagnostic diagnostic)
                 _reportDiagnostic(diagnostic);
             }
         }
+
+        /// <summary>
+        /// Adds additional file in the form of a <see cref="string"/>.  Can only be called if <see
+        /// cref="CanAddOutputFile"/> is <see langword="true"/>.
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this analyzer</param>
+        /// <param name="source">The source code to be add</param>
+        public void AddOutputFile(string hintName, string source) => AddOutputFile(hintName, SourceText.From(source, Encoding.UTF8));
```

Comment:
As we discussed offline, I don't think we should pollute the core analysis context APIs with all these new code gen specific APIs. Instead, we should take an approach similar to [DiagnosticSuppressor.ReportSuppressions](https://github.com/dotnet/roslyn/blob/d8eb54d3c56f3b8415eb713b94759bce085e13c7/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticSuppressor.cs#L29-L38) API: define a new code-gen specific sub-type of DiagnosticAnalyzer, which has code gen specific APIs and descriptors.

Reply:
Related question: why does each context need to have this API? Each context needs to have associated ReportDiagnsotic API as the diagnostic is specifically tied to the input context (for example, we use it to determine if the reported diagnostic is local or non-local + need it for partial analysis, such as single file analysis in IDE, to compute diagnostics). However, code gen scenario seems very similar to suppressor scenario, where the generated output file needs to be collected only at the end of compilation. So, we should have a single API, similar to `DiagnosticSuppressor.ReportSuppressions`, which is invoked at compilation end after all analyzers have finished executing to collect all generated output files. These code-gen analyzers can register regular analyzer callbacks to build their final output files and report them when this API is called.

Reply:
Another alternative is to have this API only on `CompilationAnalysisContext`, which is used for both CompilationEndActions and CompilationActions, though I still prefer the approach suggested in prior comment as it completely de-couples the code gen specific APIs from core analysis context APIs - if I am a regular DiagnosticAnalyzer author, I never see an API related to code gen.

Reply:
I don't quite get how this would work only on CompilationAnalysisContext.  It woudl negate all the benefits of this being callback based.  i.e. the point here is that we can generate code (often per doc) without redoing all the work the compiler has done.  If we onl yhave ComplationAnalysisContext, we'd have to go rebind everything in the semantic moels, undoing all that benefit

Reply:
> However, code gen scenario seems very similar to suppressor scenario, where the generated output file needs to be collected only at the end of compilation.

For the primary use case, we want to run this against each method, or potentially each file.  That way we can emit our per-file lsif info, while also benefitting maximally from the compiler having bound all that information for us.  We do not want to rebind everything as that would cause the entire perf cost that we're trying to avoid in the first place.

Reply:
Ok.  have moved to a model where there is a new subclass (ArtifactGenerator) and new context objects.